test/e2e/network/traffic_distribution.go

```diff
@@ -0,0 +1,319 @@
+					logLines := reverseChronologicalLogLines
+					if len(logLines) < 20 {
+						return false, fmt.Errorf("got %d log lines, waiting for at least 10", len(logLines))
+					}
+					consecutiveSameZone := 0
+
+					for _, logLine := range logLines {
+						if logLine == "" || strings.HasPrefix(logLine, "Date:") {
+							continue
+						}
+						destZone, ok := zoneForServingPod[logLine]
+						if !ok {
+							return false, fmt.Errorf("could not determine dest zone from log line: %s", logLine)
+						}
+						if clientZone != destZone {
+							return false, fmt.Errorf("expected request from clientPod=%q to stay in it's zone=%q, delivered to zone=%q", clientPod.Name, clientZone, destZone)
+						}
+						consecutiveSameZone++
+						if consecutiveSameZone >= 10 {
+							return true, nil
+						}
+					}
+					return false, nil
+				})
+
+				gomega.Eventually(ctx, requestsFromClient(clientPod)).WithPolling(5 * time.Second).WithTimeout(e2eservice.KubeProxyLagTimeout).Should(requestsSucceedAndStayInSameZone)
+			}
+
+			ginkgo.By("routing traffic cluster-wide, when there are no serving pods in the same zone as the client")
+
+			clientZone := zones[2]
+			framework.Logf("creating a client pod for probing the service from zone=%q which DOES NOT has a serving pod", clientZone)
+			clientPod := createClientPod(ctx, clientZone)
+
+			framework.Logf("ensuring that requests from clientPod=%q on zone=%q (without a serving pod) are not dropped, and get routed to one of the serving pods anywhere in the cluster", clientPod.Name, clientZone)
+
+			requestsSucceedByReachingAnyServingPod := gcustom.MakeMatcher(func(reverseChronologicalLogLines []string) (bool, error) {
+				logLines := reverseChronologicalLogLines
+				if len(logLines) < 20 {
+					return false, fmt.Errorf("got %d log lines, waiting for at least 10", len(logLines))
+				}
+
+				// Requests are counted as successful when the response read from the log
+				// lines is the name of a recognizable serving pod.
+				consecutiveSuccessfulRequests := 0
+
+				for _, logLine := range logLines {
+					if logLine == "" || strings.HasPrefix(logLine, "Date:") {
+						continue
+					}
+					_, servingPodExists := zoneForServingPod[logLine]
+					if !servingPodExists {
+						return false, fmt.Errorf("request from client pod likely failed because we got an unrecognizable response = %v; want response to be one of the serving pod names", logLine)
+					}
+					consecutiveSuccessfulRequests++
+					if consecutiveSuccessfulRequests >= 10 {
+						return true, nil
+					}
+				}
+				return false, nil
+			})
+
+			gomega.Eventually(ctx, requestsFromClient(clientPod)).WithPolling(5 * time.Second).WithTimeout(e2eservice.KubeProxyLagTimeout).Should(requestsSucceedByReachingAnyServingPod)
+
+		})
+
+	})
+
+	ginkgo.When("Service initially has trafficDistribution=PreferClose and transitions to annotation service.kubernetes.io/topology-mode=Auto and back", func() {
```

Comment:
This is an integration test, check https://github.com/kubernetes/kubernetes/tree/master/test/integration/endpointslice or https://github.com/kubernetes/kubernetes/blob/master/test/integration/service/service_test.go as examples

Reply:
Thanks. Rewrote as an integration-test.